src/harness/reference_models/fss_purge/fss_purge.py

```diff
@@ -0,0 +1,333 @@
+
+    oobe_interference = mcbsd + gcbsd - lcbsd + gfss
+    cbsd['oobe_interference'] = oobe_interference
+
+def getMcbsdValue(grant_request):
+  """ Identifies the MCBSD value for a grant.
+
+  MCBSD is the CBSD conducted power (in dBm) on the frequency segment, 
+  using the Tx mask of CBSD according to R0-DEV-05(e).
+
+  Args:
+    grant_request: A GrantData object.
+  Returns:
+    MCBSD value
+  """
+  if grant_request['operationParam']['operationFrequencyRange']['highFrequency'] > 3690000000:
+      mcbsd = -13
+  else:
+      mcbsd = -25
+
+  return mcbsd
+
+def getAntennaGainTowardsFss(incidence_angle, cbsd):
+  """Get antenna gain of CBSD in direction of FSS protection point [R2-SGN-20]
+
+  Args:
+    incidence_angle: Incidence angle as identified by ITM model.
+    cbsd: A CbsdData object
+  Returns
+    Antenna gain towards FSS in dBi.
+  """
+
+  cbsd_antenna_gain = antenna.GetStandardAntennaGains( incidence_angle.hor_cbsd,\
+                       cbsd['registrationRequest']['installationParam']['antennaAzimuth'],
+                       cbsd['registrationRequest']['installationParam']['antennaBeamwidth'],
+                       cbsd['registrationRequest']['installationParam']['antennaGain'])
+
+  return cbsd_antenna_gain
+
+def getMeanPathLossFromCbsdToFss(cbsd, fss):
+  """Get mean ITM path loss from CBSD to FSS protection point.
+
+  Args:
+    cbsd: A CbsdData object
+    fss_entity: The FssProtectionPoint tuple of the FSS.
+  Returns:
+    Mean path loss from CBSD to FSS in dBi.
+  """
+
+  # Extract the CBSD parameters
+  cbsd_latitude = cbsd['registrationRequest']['installationParam']['latitude']
+  cbsd_longitude = cbsd['registrationRequest']['installationParam']['longitude']
+  cbsd_height = cbsd['registrationRequest']['installationParam']['height']
+  cbsd_height_type = cbsd['registrationRequest']['installationParam']['heightType']
+  cbsd_indoor_dep = cbsd['registrationRequest']['installationParam']['indoorDeployment']
+  if cbsd_height_type == 'AMSL':
+      cbsd_height_agl = convertAmslToAgl(cbsd_latitude, cbsd_longitude, cbsd_height)
+  else:
+      cbsd_height_agl = cbsd_height
+
+  # Invoking ITM propagation model
+  db_loss, incidence_angle, _ = wf_itm.CalcItmPropagationLoss(
+                                         cbsd_latitude,
+                                         cbsd_longitude,
+                                         cbsd_height_agl,
+                                         fss.latitude,
+                                         fss.longitude,
+                                         fss.height_agl,
+                                         cbsd_indoor_dep,
+                                         reliability=-1,
+                                         freq_mhz=FREQ_PROP_MODEL)
+
+  return db_loss, incidence_angle
+
+def convertAmslToAgl(latitude, longitude, height_amsl):
+  """ Convert AMSL height to AGL height."""
+
+  altitude_cbsd = terrainDriver.GetTerrainElevation(latitude, longitude)

```

Comment:
This can still not work, terrainDriver is not define in this workspace.
Use instead: ```wf_itm.terrainDriver.Get...```

Note that this should have been captured if you add some basic unit test. I would recommend that you add some unit test, in the same fashion that we find in other parts of the reference_models. 
Basically a file named fss_purge_test.py, which try to unit test your interface on some simple scenarios.
To be independent of the propag results, we need to mock the propagation routines. 
One simple method is to introduce a 'fake' ITM propagation model that returns a pathloss equal to the distance + an offset.  Hence we can build scenarios, and test basic functioning.
At the very least we should have an 'example' like was done for the DPA and PPA reference models, which exercise the code, and we rely on internal testing performed by the various companies.

Reply:
Thanks. The terrain driver is working after making the change you suggested. We had some testing done on this model but all the height values were in AGL format hence this piece of code got missed. 

We have added new ‘testdata’ directory containing CbsdData objects used for unit testing of FSS Purge Model. A new file containing the unittest code has also been added. 

Now we have made the changes and tested it. We have tested the model by exercising the code similar to DPA and PPA reference models. Mocking was required at many places for db_loss, incidence_angles, antenna_gain etc.  As the main function of this model is to ensure that the sorted list (sfss) and remove list (pfss) are handled correctly and the grants are removed properly. We didn’t go for ‘mock’ approach. We fine-tuned the test data in order to hit the below cases
1.	More than one iteration is done
2.	A CBSD has some grant removed and some grants remaining
3.	A CBSD has all grants removed and hence is removed from the input FAD objects list
4.	A CBSD in pfss gets added to sfss list and causes another CBSD to move into pfss list.

Test result
Input list:  cbsd_1_grant_1,cbsd_1_grant_2,cbsd_1_grant_3,cbsd_2_grant_1,cbsd_2_grant_2,cbsd_2_grant_3,cbsd_3_grant_1,cbsd_3_grant_2,cbsd_3_grant_3,cbsd_4_grant_1,cbsd_5_grant_1,cbsd_6_grant_1
Output list: cbsd_1_grant_1,cbsd_1_grant_2,cbsd_3_grant_1,cbsd_3_grant_2,cbsd_3_grant_3,cbsd_4_grant_1,cbsd_5_grant_1,cbsd_6_grant_1

Reply:
Thanks for adding some unit test. This is a good direction to take as it will give us confidence we do not introduce regression.
Regarding the approach for unit testing and mocking, I have just proposed a FakePropagation class in another PR #476, which allow for having a controlled propagation, so that unit test  are more easy to create (for example please check the ppa_test.py file I created on that PR). 
I will continue to add utility libraries as I see fit to help doing such unit testing.